File: 11_动态规划/背包问题/4_分组背包/习题/2218. 从栈中取出 K 个硬币的最大面值和-分组背包dp.ts
// 2218. 从栈中取出K个硬币的最大面值和
// https://leetcode.cn/problems/maximum-value-of-k-coins-from-piles/
// 1 <= k <= sum(piles[i].length) <= 2000
// 1 <= n <= 1000
// 这道题时间复杂度为O(k*sum(piles[i].length))
// 分组背包，每组只能取一个前缀

function maxValueOfCoins(piles: number[][], k: number): number {
  const n = piles.length
  const dp = new Uint32Array(k + 1)
  let curCount = 0
  for (let i = 0; i < n; i++) {
    const pile = piles[i]
    const m = pile.length
    // pile 前缀和
    for (let j = 1; j < m; j++) pile[j] += pile[j - 1]
    curCount = Math.min(curCount + m, k) // 优化：j 从前 i 个栈的大小之和开始枚举（不超过 k）
    for (let j = curCount; j > 0; j--) {
      let max = 0
      for (let w = 0; w < Math.min(m, j); w++) {
        max = Math.max(max, dp[j - w - 1] + pile[w]) // w 从 0 开始，物品体积为 w+1
      }
      dp[j] = Math.max(dp[j], max)
    }
  }

  return dp[k]
}

export {}

File: 11_动态规划/背包问题/4_分组背包/习题/2218. 从栈中取出 K 个硬币的最大面值和-分组背包dp.go
package main

// https://leetcode.cn/problems/maximum-value-of-k-coins-from-piles/description/
func maxValueOfCoins(piles [][]int, k int) int {
	n := len(piles)
	dp := make([]int, k+1)
	curCount := 0
	for i := 0; i < n; i++ {
		pile := piles[i]
		m := len(pile)
		for j := 1; j < m; j++ {
			pile[j] += pile[j-1]
		}
		curCount = min(k, curCount+m)
		for j := curCount; j > 0; j-- {
			max_ := 0
			for w := 0; w < min(m, j); w++ {
				max_ = max(max_, dp[j-w-1]+pile[w])
			}
			dp[j] = max(dp[j], max_)
		}
	}
	return dp[k]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
